File: commands/update-result.js
import { SlashCommandBuilder, EmbedBuilder, MessageFlags } from 'discord.js';
import { updateMatchResult, readPlayers, readMatchVotes, readTournamentData, readCurses } from '../utils/firebase.js';
import { calculateMatches, updateGroupStandings, CURSE_PTS } from '../utils/football.js';
import { getWinner, VND_FORMATTER } from '../utils/helper.js';
import { withErrorHandler } from '../utils/command.js';
import { pickLine } from '../utils/flavor.js';
import logger from '../utils/logger.js';

export const data = new SlashCommandBuilder()
  .setName('update-result')
  .setDescription('Cập nhật kết quả trận đấu — quyền sinh sát trong tay')
  .addIntegerOption(option => option.setName('match-id')
    .setDescription('ID trận')
    .setRequired(true))
  .addIntegerOption(option => option.setName('home-score')
    .setDescription('Bàn đội nhà')
    .setMinValue(0)
    .setMaxValue(99)
    .setRequired(true))
  .addIntegerOption(option => option.setName('away-score')
    .setDescription('Bàn đội khách')
    .setMinValue(0)
    .setMaxValue(99)
    .setRequired(true));

const ALLOWED_USERS = new Set((process.env.AUDITED_USERS || '').split(',').filter(Boolean));

export const execute = withErrorHandler(async (interaction) => {
  if (!ALLOWED_USERS.has(interaction.user.id)) {
    await interaction.reply({ content: '❌ Bạn không có quyền cập nhật kết quả trận đấu.', flags: MessageFlags.Ephemeral });
    return;
  }

  await interaction.deferReply();

  const matchId = interaction.options.get('match-id').value - 1;
  const homeScore = interaction.options.get('home-score').value;
  const awayScore = interaction.options.get('away-score').value;

  const allMatches = await readTournamentData('matches');
  const matchData = allMatches?.[matchId];
  if (matchData && !matchData.messageId) {
    const embed = new EmbedBuilder()
      .setTitle('⚠️  Trận chưa được đăng')
      .setDescription(`Trận \`#${matchId + 1}\` chưa được đăng lên channel — chưa ai vote được thì cập nhật kết quả làm gì?`)
      .setColor(0xFEE75C)
      .setTimestamp();
    await interaction.editReply({ embeds: [embed] });
    return;
  }

  if (matchData?.date) {
    const kickoff = Date.parse(matchData.date);
    const elapsed = Date.now() - kickoff;
    const MIN_90 = 90 * 60 * 1000;
    if (elapsed < MIN_90) {
      const remaining = Math.ceil((MIN_90 - elapsed) / 60000);
      const embed = new EmbedBuilder()
        .setTitle('⏳  Chưa đủ 90 phút')
        .setDescription(`Trận \`#${matchId + 1}\` mới đá được chút xíu — chờ thêm **${remaining} phút** nữa rồi hãy cập nhật.`)
        .setColor(0xFEE75C)
        .setTimestamp();
      await interaction.editReply({ embeds: [embed] });
      return;
    }
  }

  const result = await updateMatchResult(matchId, homeScore, awayScore);

  if (!result.success) {
    const embed = new EmbedBuilder().setTimestamp();

    if (result.error === 'not_found') {
      embed
        .setTitle('❌  Không tìm thấy trận')
        .setDescription(`Không tìm thấy trận đấu với ID \`${matchId + 1}\`.`)
        .setColor(0xED4245);
    } else if (result.error === 'already_exists') {
      const m = result.match;
      embed
        .setTitle('⚠️  Kết quả đã tồn tại')
        .setDescription(`**${m.home.toUpperCase()}** ${m.result.home} - ${m.result.away} **${m.away.toUpperCase()}**`)
        .setColor(0xFEE75C);
    }

    await interaction.editReply({ embeds: [embed] });
    return;
  }

  const m = result.match;
  const embed = new EmbedBuilder()
    .setTitle('✅  Đã cập nhật kết quả')
    .setDescription(`**${m.home.toUpperCase()}** ${homeScore} - ${awayScore} **${m.away.toUpperCase()}**`)
    .setColor(0x57F287)
    .addFields(
      { name: '🏟️ Sân vận động', value: m.location, inline: true },
      { name: '🆔 Match ID', value: `${matchId + 1}`, inline: true },
    )
    .setFooter({ text: `Cập nhật bởi ${interaction.user.displayName}` })
    .setTimestamp();

  await interaction.editReply({ embeds: [embed] });

  const updatedMatch = {
    ...m,
    hasResult: true,
    result: { home: homeScore, away: awayScore },
  };
  await updateGroupStandings(updatedMatch);
  const matchDeltas = await calculateMatches([updatedMatch], interaction.client);
  logger.info(`Immediate calculation triggered for match ${matchId + 1}`);

  await postMatchBreakdown(interaction, updatedMatch, matchDeltas?.[updatedMatch.id]);
  await postCurseResults(interaction, updatedMatch, matchDeltas?.[updatedMatch.id]);
  await postStandings(interaction);
  await postMatchRoast(interaction, updatedMatch);
});

async function postMatchBreakdown(interaction, match, deltas) {
  try {
    if (!deltas || Object.keys(deltas).length === 0) return;

    const users = interaction.client.cachedUsers;
    const entries = Object.entries(deltas)
      .map(([userId, d]) => ({
        name: users[userId]?.nickname || 'Unknown',
        ...d,
        delta: Math.round(d.delta * 100) / 100,
      }))
      .sort((a, b) => b.delta - a.delta);

    const winner = getWinner(match);
    const allWin = entries.every((e) => e.isWinner);
    const allLose = entries.every((e) => !e.isWinner);

    const lines = entries.map((e) => {
      const sign = e.delta >= 0 ? '+' : '';
      const icon = e.isWinner ? '👑' : '🤡';
      const tag = e.random ? (e.usedRandom ? ' 🎲' : ' 🤖') : '';
      return `${icon} **${e.name}** — ${e.pick.toUpperCase()}${tag} → **${sign}${e.delta}** pts`;
    });

    if (allWin) {
      lines.push('', await pickLine('all_win'));
    } else if (allLose) {
      lines.push('', await pickLine('all_lose'));
    }

    const embed = new EmbedBuilder()
      .setTitle(`💰  Sổ sách trận #${match.id}`)
      .setDescription(lines.join('\n'))
      .setColor(allWin ? 0x57F287 : allLose ? 0xED4245 : 0x5865F2)
      .setTimestamp();

    await interaction.followUp({ embeds: [embed] });
  } catch (err) {
    logger.error('Failed to post match breakdown:', err);
  }
}

async function postCurseResults(interaction, match, deltas) {
  try {
    const curses = await readCurses();
    const matchCurses = curses[match.id];
    if (!matchCurses || Object.keys(matchCurses).length === 0) return;

    const winner = getWinner(match);
    if (!winner) return;

    const users = interaction.client.cachedUsers;

    const lines = [];
    for (const [curserId, { target }] of Object.entries(matchCurses)) {
      const curserName = users[curserId]?.nickname || 'Unknown';
      const targetName = users[target]?.nickname || 'Unknown';

      const targetPick = deltas?.[target]?.pick ?? null;
      if (targetPick === null) continue;

      const targetCorrect = targetPick === winner;
      const autoTag = deltas?.[target]?.random
        ? (deltas[target].usedRandom ? ' *(random)*' : ' *(auto)*')
        : '';
      if (targetCorrect) {
        lines.push(`🧿 **${curserName}** nguyền **${targetName}**${autoTag} — người đó đúng! **${curserName}** mất **${CURSE_PTS}** pts. ${await pickLine('curse_lose')}`);
      } else {
        lines.push(`🧿 **${curserName}** nguyền **${targetName}**${autoTag} — người đó sai! **${curserName}** ăn **${CURSE_PTS}** pts. ${await pickLine('curse_win')}`);
      }
    }

    if (lines.length === 0) return;

    const embed = new EmbedBuilder()
      .setTitle(`🧿  Bùa chú trận #${match.id}`)
      .setDescription(lines.join('\n'))
      .setColor(0x9B59B6)
      .setTimestamp();

    await interaction.followUp({ embeds: [embed] });
  } catch (err) {
    logger.error('Failed to post curse results:', err);
  }
}

async function postMatchRoast(interaction, match) {
  try {
    if (!match.messageId) return;
    const votes = await readMatchVotes(match.id, match.messageId);
    if (!votes) return;

    const winner = getWinner(match);
    const users = interaction.client.cachedUsers;
    const losers = [];

    for (const [userId, v] of Object.entries(votes)) {
      if (v.vote !== winner) {
        losers.push(users[userId]?.nickname || 'Unknown');
      }
    }

    if (losers.length === 0) return;

    const roasted = losers.slice(0, 3);
    const lines = [];
    for (const name of roasted) {
      lines.push(`🤡 **${name}** ${await pickLine('roast')}`);
    }

    if (losers.length > 3) {
      lines.push(`...cùng **${losers.length - 3}** thánh sai khác`);
    }

    const embed = new EmbedBuilder()
      .setTitle('🔥  Xào nát sau trận')
      .setDescription(lines.join('\n'))
      .setColor(0xE67E22)
      .setTimestamp();

    await interaction.followUp({ embeds: [embed] });
  } catch (err) {
    logger.error('Failed to post match roast:', err);
  }
}

async function postStandings(interaction) {
  try {
    const players = await readPlayers();
    if (!players) return;
    const users = interaction.client.cachedUsers;

    const ranked = [];
    for (const [key, value] of Object.entries(players)) {
      ranked.push({
        id: key,
        nickname: users[key]?.nickname || 'Unknown',
        points: value.points,
      });
    }
    ranked.sort((a, b) => b.points - a.points);

    if (ranked.length < 2) return;

    const leader = ranked[0];
    if (leader.points <= 0) return;

    const bottom = ranked[ranked.length - 1];

    const leaderLine = await pickLine('leader');
    const bottomLine = await pickLine('bottom');

    const lines = [
      `👑 **${leader.nickname}** ${leaderLine} (${VND_FORMATTER.format(leader.points * 1000)})`,
      '',
      `💀 **${bottom.nickname}** ${bottomLine} (${VND_FORMATTER.format(bottom.points * 1000)})`,
    ];

    const embed = new EmbedBuilder()
      .setTitle('📊  BXH sau trận')
      .setDescription(lines.join('\n'))
      .setColor(0xFFD700)
      .setTimestamp();

    await interaction.followUp({ embeds: [embed] });
  } catch (err) {
    logger.error('Failed to post standings:', err);
  }
}

File: commands/update_result.go
package commands

import (
	"context"
	"fmt"
	"log/slog"
	"os"
	"strings"
	"time"

	"github.com/kdnvi/rambo-bot/bot"
	"github.com/kdnvi/rambo-bot/internal/badges"
	"github.com/kdnvi/rambo-bot/internal/discord"
	"github.com/kdnvi/rambo-bot/internal/firebase"
	"github.com/kdnvi/rambo-bot/internal/flavor"
	"github.com/kdnvi/rambo-bot/internal/football"
)

func UpdateResult(ctx context.Context, b *bot.Bot, i *discord.Interaction) {
	allowed := allowedUsers()
	if !allowed[i.ActingUser().ID] {
		replyErr(ctx, b, i, "❌ Bạn không có quyền cập nhật kết quả trận đấu.")
		return
	}

	matchID := int(i.Data.GetOption("match-id").Int())
	homeScore := int(i.Data.GetOption("home-score").Int())
	awayScore := int(i.Data.GetOption("away-score").Int())
	matchIndex := matchID - 1

	if err := deferReply(ctx, b, i, false); err != nil {
		return
	}

	matches, err := b.DB.ReadMatches(ctx)
	if err != nil {
		editReply(ctx, b, i, discord.EditMessagePayload{Content: "❌ Không có dữ liệu trận đấu."})
		return
	}

	if matchIndex < 0 || matchIndex >= len(matches) {
		editReply(ctx, b, i, discord.EditMessagePayload{
			Embeds: []discord.Embed{{
				Title:       "❌  Không tìm thấy trận",
				Description: fmt.Sprintf("Không tìm thấy trận đấu với ID `%d`.", matchID),
				Color:       0xED4245,
				Timestamp:   time.Now().UTC().Format(time.RFC3339),
			}},
		})
		return
	}

	matchData := matches[matchIndex]

	if matchData.MessageID == "" {
		editReply(ctx, b, i, discord.EditMessagePayload{
			Embeds: []discord.Embed{{
				Title:       "⚠️  Trận chưa được đăng",
				Description: fmt.Sprintf("Trận `#%d` chưa được đăng lên channel — chưa ai vote được thì cập nhật kết quả làm gì?", matchID),
				Color:       0xFEE75C,
				Timestamp:   time.Now().UTC().Format(time.RFC3339),
			}},
		})
		return
	}

	if matchData.Date != "" {
		kickoff, _ := time.Parse(time.RFC3339, matchData.Date)
		elapsed := time.Since(kickoff)
		if elapsed < 90*time.Minute {
			remaining := int((90*time.Minute - elapsed).Minutes())
			editReply(ctx, b, i, discord.EditMessagePayload{
				Embeds: []discord.Embed{{
					Title:       "⏳  Chưa đủ 90 phút",
					Description: fmt.Sprintf("Trận `#%d` mới đá được chút xíu — chờ thêm **%d phút** nữa rồi hãy cập nhật.", matchID, remaining),
					Color:       0xFEE75C,
					Timestamp:   time.Now().UTC().Format(time.RFC3339),
				}},
			})
			return
		}
	}

	prevMatch, err := b.DB.UpdateMatchResult(ctx, matchIndex, homeScore, awayScore)
	if err != nil {
		if err.Error() == "not_found" {
			editReply(ctx, b, i, discord.EditMessagePayload{
				Embeds: []discord.Embed{{
					Title:       "❌  Không tìm thấy trận",
					Description: fmt.Sprintf("Không tìm thấy trận đấu với ID `%d`.", matchID),
					Color:       0xED4245,
					Timestamp:   time.Now().UTC().Format(time.RFC3339),
				}},
			})
		} else if err.Error() == "already_exists" {
			editReply(ctx, b, i, discord.EditMessagePayload{
				Embeds: []discord.Embed{{
					Title:       "⚠️  Kết quả đã tồn tại",
					Description: fmt.Sprintf("**%s** %d - %d **%s**", upper(prevMatch.Home), prevMatch.Result.Home, prevMatch.Result.Away, upper(prevMatch.Away)),
					Color:       0xFEE75C,
					Timestamp:   time.Now().UTC().Format(time.RFC3339),
				}},
			})
		} else {
			editReply(ctx, b, i, discord.EditMessagePayload{Content: "❌ Có lỗi xảy ra."})
		}
		return
	}

	editReply(ctx, b, i, discord.EditMessagePayload{
		Embeds: []discord.Embed{{
			Title:       "✅  Đã cập nhật kết quả",
			Description: fmt.Sprintf("**%s** %d - %d **%s**", upper(prevMatch.Home), homeScore, awayScore, upper(prevMatch.Away)),
			Color:       0x57F287,
			Fields: []discord.EmbedField{
				{Name: "🏟️ Sân vận động", Value: prevMatch.Location, Inline: true},
				{Name: "🆔 Match ID", Value: fmt.Sprintf("%d", matchID), Inline: true},
			},
			Footer:    &discord.EmbedFooter{Text: fmt.Sprintf("Cập nhật bởi %s", i.ActingUser().DisplayName())},
			Timestamp: time.Now().UTC().Format(time.RFC3339),
		}},
	})

	// Build the updated match for calculations
	updatedMatch := *prevMatch
	updatedMatch.HasResult = true
	updatedMatch.Result = &firebase.MatchResult{Home: homeScore, Away: awayScore}

	// Update group standings (group stage only)
	if err := football.UpdateGroupStandings(ctx, b.DB, &updatedMatch); err != nil {
		slog.Error("update group standings failed", "err", err)
	}

	// Calculate points
	matchDeltas, err := football.CalculateMatches(ctx, b.DB, []firebase.Match{updatedMatch})
	if err != nil {
		slog.Error("calculate matches failed", "err", err)
	}
	slog.Info("immediate calculation triggered", "match", matchID)

	// Check badges
	if matchDeltas != nil {
		allMatches, _ := b.DB.ReadMatches(ctx)
		completed := make([]firebase.Match, 0)
		for _, m := range allMatches {
			if m.HasResult && m.IsCalculated {
				completed = append(completed, m)
			}
		}
		badges.CheckAndAward(ctx, b, completed, updatedMatch.ID)
	}

	// Post breakdown, curse results, standings, roast
	postMatchBreakdown(ctx, b, i, &updatedMatch, matchDeltas[updatedMatch.ID])
	postCurseResults(ctx, b, i, &updatedMatch, matchDeltas[updatedMatch.ID])
	postStandings(ctx, b, i)
	postMatchRoast(ctx, b, i, &updatedMatch)
}

func postMatchBreakdown(ctx context.Context, b *bot.Bot, i *discord.Interaction, match *firebase.Match, deltas map[string]*football.Delta) {
	if len(deltas) == 0 {
		return
	}
	users := b.GetAllCachedUsers()

	type entry struct {
		name     string
		delta    float64
		pick     string
		isWinner bool
		random   bool
	}
	entries := make([]entry, 0, len(deltas))
	for uid, d := range deltas {
		name := "Unknown"
		if u := users[uid]; u != nil {
			name = u.DisplayName()
		}
		entries = append(entries, entry{name, round2(d.Delta), d.Pick, d.IsWinner, d.Random})
	}
	// Sort winners first, then by delta desc
	for a := 0; a < len(entries); a++ {
		for b2 := a + 1; b2 < len(entries); b2++ {
			if entries[b2].delta > entries[a].delta {
				entries[a], entries[b2] = entries[b2], entries[a]
			}
		}
	}

	allWin := true
	allLose := true
	for _, e := range entries {
		if !e.isWinner {
			allWin = false
		}
		if e.isWinner {
			allLose = false
		}
	}

	lines := make([]string, 0, len(entries))
	for _, e := range entries {
		sign := ""
		if e.delta >= 0 {
			sign = "+"
		}
		icon := "🤡"
		if e.isWinner {
			icon = "👑"
		}
		tag := ""
		if e.random {
			tag = " 🎲"
		}
		lines = append(lines, fmt.Sprintf("%s **%s** — %s%s → **%s%.2f** pts", icon, e.name, upper(e.pick), tag, sign, e.delta))
	}

	if allWin {
		lines = append(lines, "", flavor.PickLine(ctx, b.DB, "all_win"))
	} else if allLose {
		lines = append(lines, "", flavor.PickLine(ctx, b.DB, "all_lose"))
	}

	color := 0x5865F2
	if allWin {
		color = 0x57F287
	} else if allLose {
		color = 0xED4245
	}

	followup(ctx, b, i, discord.SendMessagePayload{
		Embeds: []discord.Embed{{
			Title:       fmt.Sprintf("💰  Sổ sách trận #%d", match.ID),
			Description: strings.Join(lines, "\n"),
			Color:       color,
			Timestamp:   time.Now().UTC().Format(time.RFC3339),
		}},
	})
}

func postCurseResults(ctx context.Context, b *bot.Bot, i *discord.Interaction, match *firebase.Match, deltas map[string]*football.Delta) {
	curses, err := b.DB.ReadCurses(ctx)
	if err != nil {
		return
	}
	matchCurses := curses[fmt.Sprintf("%d", match.ID)]
	if len(matchCurses) == 0 {
		return
	}
	winner := football.GetWinner(match)
	if winner == "" {
		return
	}

	users := b.GetAllCachedUsers()
	lines := []string{}
	for curserID, curse := range matchCurses {
		curserName := "Unknown"
		if u := users[curserID]; u != nil {
			curserName = u.DisplayName()
		}
		targetName := "Unknown"
		if u := users[curse.Target]; u != nil {
			targetName = u.DisplayName()
		}
		d := deltas[curse.Target]
		if d == nil {
			continue
		}
		autoTag := ""
		if d.Random {
			autoTag = " *(auto)*"
		}
		targetCorrect := d.Pick == winner
		if targetCorrect {
			lines = append(lines, fmt.Sprintf("🧿 **%s** nguyền **%s**%s — người đó đúng! **%s** mất **%d** pts. %s",
				curserName, targetName, autoTag, curserName, football.CursePts, flavor.PickLine(ctx, b.DB, "curse_lose")))
		} else {
			lines = append(lines, fmt.Sprintf("🧿 **%s** nguyền **%s**%s — người đó sai! **%s** ăn **%d** pts. %s",
				curserName, targetName, autoTag, curserName, football.CursePts, flavor.PickLine(ctx, b.DB, "curse_win")))
		}
	}
	if len(lines) == 0 {
		return
	}
	followup(ctx, b, i, discord.SendMessagePayload{
		Embeds: []discord.Embed{{
			Title:       fmt.Sprintf("🧿  Bùa chú trận #%d", match.ID),
			Description: strings.Join(lines, "\n"),
			Color:       0x9B59B6,
			Timestamp:   time.Now().UTC().Format(time.RFC3339),
		}},
	})
}

func postMatchRoast(ctx context.Context, b *bot.Bot, i *discord.Interaction, match *firebase.Match) {
	if match.MessageID == "" {
		return
	}
	votes, err := b.DB.ReadMatchVotes(ctx, match.ID, match.MessageID)
	if err != nil || len(votes) == 0 {
		return
	}
	winner := football.GetWinner(match)
	users := b.GetAllCachedUsers()

	var losers []string
	for uid, v := range votes {
		if v.Vote != winner {
			name := "Unknown"
			if u := users[uid]; u != nil {
				name = u.DisplayName()
			}
			losers = append(losers, name)
		}
	}
	if len(losers) == 0 {
		return
	}

	roasted := losers
	if len(roasted) > 3 {
		roasted = roasted[:3]
	}
	lines := make([]string, 0, len(roasted))
	for _, name := range roasted {
		lines = append(lines, fmt.Sprintf("🤡 **%s** %s", name, flavor.PickLine(ctx, b.DB, "roast")))
	}
	if len(losers) > 3 {
		lines = append(lines, fmt.Sprintf("...cùng **%d** thánh sai khác", len(losers)-3))
	}

	followup(ctx, b, i, discord.SendMessagePayload{
		Embeds: []discord.Embed{{
			Title:       "🔥  Xào nát sau trận",
			Description: strings.Join(lines, "\n"),
			Color:       0xE67E22,
			Timestamp:   time.Now().UTC().Format(time.RFC3339),
		}},
	})
}

func postStandings(ctx context.Context, b *bot.Bot, i *discord.Interaction) {
	players, err := b.DB.ReadPlayers(ctx)
	if err != nil || len(players) < 2 {
		return
	}
	users := b.GetAllCachedUsers()

	type entry struct {
		id     string
		name   string
		points float64
	}
	ranked := make([]entry, 0, len(players))
	for id, p := range players {
		name := "Unknown"
		if u := users[id]; u != nil {
			name = u.DisplayName()
		}
		ranked = append(ranked, entry{id, name, p.Points})
	}
	for a := 0; a < len(ranked); a++ {
		for b2 := a + 1; b2 < len(ranked); b2++ {
			if ranked[b2].points > ranked[a].points {
				ranked[a], ranked[b2] = ranked[b2], ranked[a]
			}
		}
	}
	if ranked[0].points <= 0 {
		return
	}

	leader := ranked[0]
	bottom := ranked[len(ranked)-1]

	leaderLine := flavor.PickLine(ctx, b.DB, "leader")
	bottomLine := flavor.PickLine(ctx, b.DB, "bottom")

	lines := []string{
		fmt.Sprintf("👑 **%s** %s (%s)", leader.name, leaderLine, fmtVND(leader.points)),
		"",
		fmt.Sprintf("💀 **%s** %s (%s)", bottom.name, bottomLine, fmtVND(bottom.points)),
	}

	followup(ctx, b, i, discord.SendMessagePayload{
		Embeds: []discord.Embed{{
			Title:       "📊  BXH sau trận",
			Description: strings.Join(lines, "\n"),
			Color:       0xFFD700,
			Timestamp:   time.Now().UTC().Format(time.RFC3339),
		}},
	})
}

func allowedUsers() map[string]bool {
	m := map[string]bool{}
	for _, id := range strings.Split(os.Getenv("AUDITED_USERS"), ",") {
		if id != "" {
			m[id] = true
		}
	}
	return m
}

func round2(v float64) float64 {
	if v < 0 {
		return float64(int64(v*100-0.5)) / 100
	}
	return float64(int64(v*100+0.5)) / 100
}
